ast/walk.go

```diff
@@ -72,6 +72,112 @@ func Walk(expr Expression, visitor func(Expression) bool) bool {
 	panic(fmt.Sprintf("Unexpected AST expression of type %T: %[1]v", expr))
 }
 
+// WalkAndFindParentUsageInvalid is like Walk() but it returns an error if an invalid
+// parent usage is found.
+// This function is exclusively used to walk the body of a custom defined function.
+// The usage of a parent operator is *not* allowed in the body of a custom defined function
+// if it reaches out of that function's scope. It's valid within a filter or a projection, because
+// we create NewNestedScope for those.
+func WalkAndFindParentUsageInvalid(count int, expr Expression) (int, error) {
+	if expr == nil {
+		return count, nil
+	}
+	switch e := expr.(type) {
+	case *Parent:
+		count--
+		if count <= 0 {
+			return count, fmt.Errorf("Parent usage is invalid")
+		}
+		return count, nil
+	case *Everything, *This, *Ellipsis, *StringLiteral,
+		*FloatLiteral, *BooleanLiteral, *IntegerLiteral, *NullLiteral, *Attribute,
+		*ArrayTraversal, *Param, *FunctionParam:
+		return count, nil
+	case *Constraint:
+		return WalkAndFindParentUsageInvalid(count, e.Expression)
+	case *Subscript:
+		return WalkAndFindParentUsageInvalid(count, e.Value)
+	case *Range:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.Start); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.End)
+	case *BinaryOperator:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.RHS)
+	case *DotOperator:
+		// Chain the count between LHS and RHS sides of the dot operator.
+		c, err := WalkAndFindParentUsageInvalid(count, e.LHS)
+		if err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(c, e.RHS)
+	case *Group:
+		return WalkAndFindParentUsageInvalid(count, e.Expression)
+	case *PipeOperator:
+		if _, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return count, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.RHS)
+	case *PrefixOperator:
+		return WalkAndFindParentUsageInvalid(count, e.RHS)
+	case *PostfixOperator:
+		return WalkAndFindParentUsageInvalid(count, e.LHS)
+	case *Array:
+	case *Object:
+		for _, expr := range e.Expressions {
+			if c, err := WalkAndFindParentUsageInvalid(count, expr); err != nil {
+				return c, err
+			}
+		}
+		return count, nil
+	case *FunctionCall:
+		for _, arg := range e.Arguments {
+			if c, err := WalkAndFindParentUsageInvalid(count, arg); err != nil {
+				return c, err
+			}
+		}
+		return count, nil
+	case *FunctionPipe:
+		if c, err := WalkAndFindParentUsageInvalid(count, e.LHS); err != nil {
+			return c, err
+		}
+		return WalkAndFindParentUsageInvalid(count, e.Func)
```

Comment:
Here we should also handle `| order(…)`, no? I believe this code now incorrectly rejects `|order(^.foo)` because it doesn't know that it introduces a new scope.